compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -3059,30 +3128,83 @@ object Types {
     def derivedAppliedType(tycon: Type, args: List[Type])(implicit ctx: Context): Type =
       if ((tycon eq this.tycon) && (args eq this.args)) this
       else tycon.appliedTo(args)
+  }
 
-    override def computeHash = doHash(tycon, args)
+  final class CachedAppliedType(tycon: Type, args: List[Type], hc: Int) extends AppliedType(tycon, args) {
+    myHash = hc
+    override def computeHash = unsupported("computeHash")
+  }
 
-    protected def checkInst(implicit ctx: Context): this.type = {
-      def check(tycon: Type): Unit = tycon.stripTypeVar match {
-        case tycon: TypeRef if !tycon.symbol.isClass =>
-        case _: TypeParamRef | _: ErrorType | _: WildcardType =>
-        case _: TypeLambda =>
-          assert(!args.exists(_.isInstanceOf[TypeBounds]), s"unreduced type apply: $this")
-        case tycon: AnnotatedType =>
-          check(tycon.underlying)
-        case _ =>
-          assert(false, s"illegal type constructor in $this")
+  object AppliedType {
+    def apply(tycon: Type, args: List[Type])(implicit ctx: Context) = {
+      assertUnerased()
+      ctx.base.uniqueAppliedTypes.enterIfNew(tycon, args)
+    }
+  }
+
+  object ClassRef {
+    def unapply(tp: RefType)(implicit ctx: Context): Option[RefType] = { // after bootstrap, drop the Option
+      if (tp.symbol.isClass) Some(tp) else None
+    }
+  }
+
+  /** A reference to wildcard argument `p.<parameter X of class C>`
+   *  where `p: C[... _ ...]`
+   */
+  abstract case class TypeArgRef(prefix: Type, clsRef: TypeRef, idx: Int) extends CachedProxyType with ValueType {
```

Comment:
Complication: New type TypeArgRef. 

There's hope we can get rid of this again if we permit references to private type parameters as  general types.